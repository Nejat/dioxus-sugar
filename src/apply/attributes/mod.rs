#![allow(dead_code)]

use proc_macro2::Ident;
use std::cmp::Ordering;

mod parse;
mod tokens;

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Property {
    Optional(String),
    Required(String),
}

impl PartialOrd<Self> for Property {
    #[inline]
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Property {
    fn cmp(&self, other: &Self) -> Ordering {
        let self_value = match self { Self::Optional(value) | Self::Required(value) => value };
        let other_value = match other { Self::Optional(value) | Self::Required(value) => value };

        self_value.cmp(other_value)
    }
}

impl Property {
    pub const fn remap(&self, new_value: String) -> Self {
        match self {
            Self::Optional(_) => Self::Optional(new_value),
            Self::Required(_) => Self::Required(new_value),
        }
    }
}

impl AsRef<str> for Property {
    fn as_ref(&self) -> &str {
        match self {
            Self::Optional(value) |
            Self::Required(value) => value.as_str(),
        }
    }
}

///
pub struct ApplyAttributes {
    pub(crate) element: Ident,

    ///
    pub(crate) attributes: Vec<Property>,
}
