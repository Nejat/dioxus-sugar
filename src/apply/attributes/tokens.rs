use proc_macro2::{Ident, Span, TokenStream};
use quote::ToTokens;

use crate::ApplyAttributes;

impl ToTokens for ApplyAttributes {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let cx = &self.context;

        for attribute in &self.attributes {
            let attr = Ident::new(attribute.as_str(), Span::call_site());
            let value = quote! { {#cx.props.#attr} }.to_string();

            tokens.extend(quote! { #attr: #value, });
        }
    }
}
