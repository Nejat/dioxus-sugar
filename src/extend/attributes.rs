use std::collections::HashMap;

use proc_macro2::{Ident, Span, TokenStream};
use quote::ToTokens;
use syn::{AttributeArgs, ItemStruct, Lifetime, Path};
use web_reference::prelude::*;

use crate::extend::{extend_input_struct, Extension, net_extensions, parse_extensions_args};
use crate::SPECS;

// todo: optional attribute properties

///
pub fn input_struct(input: &mut ItemStruct, args: &AttributeArgs) -> TokenStream {
    let extensions = extract_attribute_extensions(args);

    extend_input_struct(input, &extensions, write_attribute_extension)
}

///
fn attributes_of_tag(tag: &Tag) -> Vec<&Attribute> {
    let attributes = tag.attributes.iter()
        .chain(&tag.optional_attributes)
        .filter_map(|attr| SPECS.get_attribute(attr.as_str()));

    if tag.global_attributes {
        attributes.chain(
            SPECS.get_attributes_of_category(AttributeCategory::GlobalAttributes)
                .unwrap().iter().copied()
        ).collect::<Vec<_>>()
    } else {
        attributes.collect::<Vec<_>>()
    }
}

///
fn extract_attribute_extensions(args: &AttributeArgs) -> Vec<Extension> {
    let extensions = parse_extensions_args(
        args, "attribute or tag", validate_attribute_extension,
    ).flat_map(|ext| {
        let name = ext.name.to_string();


        if name == "global" {
            SPECS.get_attributes_of_category(AttributeCategory::GlobalAttributes).unwrap().into_iter()
                .filter_map(filter_attribute_key_value(&ext))
                .collect::<Vec<_>>()
        } else if SPECS.is_valid_attribute(&name) {
            vec![(format!("{}{}", ext.name, ext.exclude), ext)]
        } else if let Some(tag) = SPECS.get_tag(&name) {
            attributes_of_tag(tag).into_iter()
                .filter_map(filter_attribute_key_value(&ext))
                .collect::<Vec<_>>()
        } else if let Ok(tags_category) = TagCategory::try_from(name.as_str()) {
            SPECS.get_tags_of_category(tags_category).unwrap().into_iter()
                .flat_map(attributes_of_tag)
                .filter_map(filter_attribute_key_value(&ext))
                .collect::<Vec<_>>()
        } else {
            unreachable!()
        }
    }).collect::<HashMap<_, _>>().into_iter()
        .map(|(_key, value)| value)
        .collect::<Vec<_>>();

    if extensions.is_empty() {
        abort!(Span::call_site(), "no attributes or tags were listed, remove 'attributes' macro or add an attribute or tag");
    }

    net_extensions(extensions, "no net attributes or tags to extend, check exclude list")
}

///
fn filter_attribute_key_value<'a>(ext: &'a Extension) -> impl Fn(&'a Attribute) -> Option<(String, Extension)> {
    |attr: &'a Attribute|
        if attr.name.starts_with("data-") {
            None
        } else {
            Some((
                format!("{}{}", attr.name, ext.exclude),
                Extension {
                    name: Ident::new(attr.name.as_str(), ext.name.span()),
                    exclude: ext.exclude,
                }))
        }
}

///
fn write_attribute(attribute: &Attribute) -> TokenStream {
    let attr_ident = Ident::new(&attribute.name, Span::call_site());
    let ty = Ident::new(match &attribute.values {
        AttributeValue::OnOff { .. } |
        AttributeValue::Boolean { .. } => "bool",
        AttributeValue::URLList { .. } => "Vec<String>",
        _ => "String",
    }, Span::call_site());

    quote! { #attr_ident: #ty }
}

///
fn write_attribute_extension(extension: &Extension, _life_time: &Option<Lifetime>) -> TokenStream {
    let name = extension.name.to_string();

    SPECS.get_attributes(&name)
        .map_or_else(
            || abort!(Span::call_site(), format!("could not find {name:?} attribute")),
            |attrs| write_attribute(attrs.values().next().unwrap()),
        )
}

///
fn validate_attribute_extension(path: &Path) {
    const GLOBAL: &str = "global";

    let attribute = path.to_token_stream().to_string();

    if SPECS.is_valid_attribute(&attribute) ||
        SPECS.is_valid_tag(&attribute) ||
        attribute == GLOBAL ||
        TagCategory::try_from(attribute.as_str()).is_ok()
    {
        return;
    }

    abort!(&path.to_token_stream(), format!("invalid attribute or tag {attribute:?}"));
}
