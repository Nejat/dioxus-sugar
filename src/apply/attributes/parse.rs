use std::collections::HashSet;

use quote::ToTokens;
use syn::ExprPath;
use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use web_reference::prelude::*;

use crate::{ApplyAttributes, SPECS};
use crate::apply::attributes::Property;
use crate::common::attributes_of_tag;

impl Parse for ApplyAttributes {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(Self {
            element: {
                let elm = input.parse()?;

                <Token![;]>::parse(input)?;

                elm
            },
            attributes: parse_attributes(input)?,
        })
    }
}

fn parse_attributes(input: ParseStream) -> syn::Result<Vec<Property>> {
    let mut include = Vec::new();
    let mut exclude = Vec::new();
    let mut optional = Vec::new();

    loop {
        let attribute = <ExprPath>::parse(input)?;

        let attr = attribute.to_token_stream().to_string();

        match attr.as_str() {
            "exclude" => parse_list_of_attributes(&mut exclude, input, Property::Required)?,
            "optional" => parse_list_of_attributes(&mut optional, input, Property::Optional)?,
            _ => {
                valid_attribute(attr.as_str(), input)?;

                include.push(Property::Required(attr));
            }
        }

        if !(input.peek(Token![,])) { break; }

        <Token![,]>::parse(input)?;
    }

    let include = expand_all_attributes(
        include.into_iter()
            .chain(optional.into_iter())
    );

    let exclude = expand_all_attributes(exclude);

    let mut attributes = include.into_iter()
        .filter(|inc| exclude.iter().all(|ex| ex.as_ref() != inc.as_ref()))
        .collect::<Vec<_>>();

    attributes.sort();

    Ok(attributes)
}

fn parse_list_of_attributes(
    list: &mut Vec<Property>, input: ParseStream, parse: fn(String) -> Property,
) -> syn::Result<()> {
    let content;

    parenthesized!(content in input);

    let excludes: Punctuated<ExprPath, Token![,]> = content.parse_terminated(ExprPath::parse)?;

    for attribute in excludes {
        let attr = attribute.to_token_stream().to_string();

        valid_attribute(attr.as_str(), input)?;

        list.push(parse(attr));
    }

    Ok(())
}

fn expand_all_attributes(attributes: impl IntoIterator<Item=Property>) -> HashSet<Property> {
    attributes.into_iter()
        .flat_map(|attribute| {
            if attribute.as_ref() == "global" {
                SPECS.get_attributes_of_category(AttributeCategory::GlobalAttributes).unwrap().into_iter()
                    .filter_map(filter_attribute)
                    .map(|new_attr| attribute.remap(new_attr))
                    .collect::<Vec<_>>()
            } else if SPECS.is_valid_attribute(attribute.as_ref()) {
                vec![attribute]
            } else if let Some(tag) = SPECS.get_tag(attribute.as_ref()) {
                attributes_of_tag(tag).into_iter()
                    .filter_map(filter_attribute)
                    .map(|new_attr| attribute.remap(new_attr))
                    .collect::<Vec<_>>()
            } else if let Ok(tags_category) = TagCategory::try_from(attribute.as_ref()) {
                SPECS.get_tags_of_category(tags_category).unwrap().into_iter()
                    .flat_map(attributes_of_tag)
                    .filter_map(filter_attribute)
                    .map(|new_attr| attribute.remap(new_attr))
                    .collect::<Vec<_>>()
            } else {
                unreachable!()
            }
        }).collect()
}

fn filter_attribute(attr: &Attribute) -> Option<String> {
    if attr.name.starts_with("data-") {
        None
    } else {
        Some(attr.name.clone())
    }
}

#[inline]
fn valid_attribute(attr: &str, input: ParseStream) -> syn::Result<()> {
    if attr == "global" ||
        SPECS.is_valid_attribute(attr) ||
        SPECS.is_valid_tag(attr) ||
        TagCategory::try_from(attr).is_ok()
    {
        Ok(())
    } else {
        Err(input.error(format!("{attr:?} is not a valid attribute, tag or tag category")))
    }
}
