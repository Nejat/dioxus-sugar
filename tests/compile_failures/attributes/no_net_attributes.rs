use dioxus_sugar::attributes;

#[attributes(href, div, exclude(href, div, class))]
struct Props {
    name: String,
    id: String,
}

#[attributes(exclude(href, div, class))]
struct Props2 {
    name: String,
    id: String,
}

fn main() {}