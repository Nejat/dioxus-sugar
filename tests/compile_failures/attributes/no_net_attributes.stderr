error: no net attributes or tags to extend, check exclude list
 --> tests/compile_failures/attributes/no_net_attributes.rs:3:1
  |
3 | #[attributes(href, div, exclude(href, div, class))]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `attributes` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no net attributes or tags to extend, check exclude list
 --> tests/compile_failures/attributes/no_net_attributes.rs:9:1
  |
9 | #[attributes(exclude(href, div, class))]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `attributes` (in Nightly builds, run with -Z macro-backtrace for more info)
